### PR TITLE
C++: Minimal IR support for `GNUVectorType`

### PR DESCRIPTION
**Pull request by dave-bartolomeo**
Lack of support for the GCC vector extensions was causing a bunch of sanity failures in the syntax zoo. This PR adds minimal IR generation support for these types.

Added `VectorAggregateLiteral`, and factored most of `ArrayAggregateLiteral` out into the common base class `ArrayOrVectorAggregateLiteral`. I'd be happy to merge these all into `ArrayAggregateLiteral` if we don't care about the distinction.

Made a few tweaks to `TranslatedArrayExpr` to compute the element type by looking at the result type of the `ArrayExpr`, not the type of the base operand. Note that this means that for `T a[10]; a[i] = foo;`, the result of the `PointerAdd` for `a[i]` will now be `glvalue<T>`, not `T*`. This is actually more faithful to the source language, and has no semantic difference on the IR.

Added some missing `getInstructionElementSize()` overrides.

Added the new `BuiltIn` opcode, renamed the existing `BuiltInInstruction` to `BuiltInOperationInstruction`, and made any `BuiltInOperation` that we don't specifically handle translate to `BuiltIn`. `BuiltInOperationInstruction` now has a way to get the specific `BuiltInOperation`.

Added `getCanonicalQLClass()` overrides for `GNUVectorType` and `BuiltInOperation`.

Added a simple IR test for vector types.

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/ir/implementation/unaliased_ssa/Instruction.qll [hunk 1]
```diff
@@ -1831,8 +1831,29 @@ class UnreachedInstruction extends Instruction {
...
+}
+
+/**
+ * An instruction representing a built-in operation that does not have a specific opcode. The
+ * actual operation is specified by the `getBuiltInOperation()` predicate.
+ */
+class BuiltInInstruction extends BuiltInOperationInstruction {
   BuiltInInstruction() {
-    getOpcode() instanceof BuiltInOpcode
+    getOpcode() instanceof Opcode::BuiltIn
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/ir/implementation/unaliased_ssa/Instruction.qll]: Is there now no `Instruction` class for the varargs builtins?

**Comment by jbj** [comment]
I re-triggered the tests, and they show a semantic merge conflict between this PR and one of the `getCanonicalQLClass` PRs that was merged:

```
10:13:11  +ERROR: Predicate getCanonicalQLClass is already declared (/opt/jenkins/jenkins/workspace/Language-Tests/CPP/job/ql/cpp/ql/src/semmle/code/cpp/Type.qll:1071,19-38)
10:13:11  +ERROR: This predicate does not override another predicate (/opt/jenkins/jenkins/workspace/Language-Tests/CPP/job/ql/cpp/ql/src/semmle/code/cpp/Type.qll:1058,3-69)
```

File: cpp/ql/src/semmle/code/cpp/ir/implementation/unaliased_ssa/Instruction.qll (hunk 1 above)
Comment (dave-bartolomeo) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/unaliased_ssa/Instruction.qll]: What used to be called `BuiltInInstruction`, which was a base class for the varargs instructions, is now called `BuiltInOperationInstruction`. What is now called `BuiltInInstruction` is the class that specifically represents instructions with the opcode `BuiltIn`, which does not include the varargs instructions, which still have their own opcode.

**Comment by dave-bartolomeo** [comment]
Semantic conflicts fixed and ready for merge.

**Pull request merged**